README.md

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

![Bolder Flight Systems Logo](img/logo-words_75.png) &nbsp; &nbsp; ![Arduino Logo](img/arduino_logo_75.png)

# CircleBuf
This library implements a first in, first out [circular buffer](https://en.wikipedia.org/wiki/Circular_buffer), methods are provided for writing to and reading from the buffer. This library is compatible with Arduino and with CMake build systems.
   * [License](LICENSE.md)
   * [Changelog](CHANGELOG.md)
   * [Contributing guide](CONTRIBUTING.md)

# Installation

## Arduino
Use the Arduino Library Manager to install this library or clone to your Arduino/libraries folder. The library is added as:

```C++
#include "circle_buf.h"
```

An example Arduino executable is located at *examples/arduino/circle_buf_example/circle_buf_example.ino*. Teensy 3.x, 4.x, and LC devices are used for testing under Arduino and this library should be compatible with other devices.

## CMake
CMake is used to build this library, which is exported as a library target called *circle_buf*. The header is added as:

```C++
#include "circle_buf.h"
```

The library can be also be compiled stand-alone using the CMake idiom of creating a *build* directory and then, from within that directory issuing:

```
cmake ..
make
```

This will build the library, an example executable called *circle_buf_example*, and an executable for testing using the Google Test framework, called *circle_buf_test*. The example executable source file is located at *examples/cmake/circle_buf_example.cc*.

# Namespace
This library is within the namespace *bfs*

# CircleBuf
This class implements the circular buffer.

## Methods

**CircleBuf<typename T, std::size_t N>** Creates a circular buffer object, of type *T* with buffer size *N*.

```C++
/* Create a 10 byte circular buffer of uint8_t */
bfs::CircleBuf<uint8_t, 10> buff;
```

**bool Write(const T val)** Writes a single value to the buffer. Note the buffer prevents over-writing data. This method returns true on success or false on failure.

```C++
bool status = buff.Write(3);
```

**std::size_t Write(T const &ast; const data, const std::size_t len)** Writes data into the buffer given a pointer to the data and the number of elements to write. Returns the number of elements written into the buffer. Note that the buffer prevents over-writing data, so the number of elements written may be less than the number of elements to write.

```C++
uint8_t test[] = {1, 2, 3, 4, 5};
std::size_t ret = buff.Write(test, sizeof(test));
```

**bool Read(T &ast; const data)** Reads a single value from the buffer given a pointer to store the data. True is returned if a value is returned through the pointer. A value would not be returned if the buffer is empty (i.e size = 0).

```C++
uint8_t val;
bool ret = buff.Read(&val);
if (ret) {
   std::cout << std::to_string(val) << std::endl;
}
```

**std::size_t Read(T &ast; const data, std::size_t len)** Reads elements from the buffer given a pointer to store the data and the maximum number of elements to read. Returns the number of elements read off the buffer.

```C++
uint8_t read_val[20];
std::size_t ret = buff.Read(read_val, sizeof(read_val));
```

**std::size_t capacity()** Returns the capacity of the circular buffer. This will always be equal to *N* used during construction.

```C++
std::cout << buff.capacity() << std::endl;
```

**std::size_t size()** Returns the number of elements currently stored in the buffer.

```C++
std::cout << buff.size() << std::endl;
```

**void Clear()** Empties the buffer.

```C++
buff.Clear();
```


CHANGELOG.md
# Changelog

## v5.0.0
- Removed optional

## v4.0.4
- Updated include guards

## v4.0.3
- Added std::optional support for the read method

## v4.0.2
- Updated CMake minimum version to 3.14 to support FetchContent_MakeAvailable

## v4.0.1
- Fixed constness of Write method

## v4.0.0
- Updated the name to CircleBuf
- Updated the includes to more accurately capture those needed

## v3.0.2
- Updated preprocessor directives to more easily define when Arduino IDE is used
- Updated README to better detail the Arduino support

## v3.0.1
- Tweaking preprocessor directives to better catch when Arduino IDE is used to compile and when it is not

## v3.0.0
- Updated directory structure to support Arduino in addition to CMake builds
- Added an Arduino example
- Updated README

## v2.0.0
- Updated to namespace *bfs*

## v1.2.1
- Updated CONTRIBUTING
- Updated CI/CD pipeline to no longer ignore line length limits
- Modified circle_buf.h to respect the line length limits

## v1.2.0
- Added method to clear the buffer

## v1.1.2
- Updated CONTRIBUTING

## v1.1.1
- Updated license to MIT

## v1.1.0
- Templated the container type
- Added single element read and write methods

## v1.0.0
- Initial baseline


tests/circle_buf_test.cc
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2023 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#include "circle_buf.h"
#include "gtest/gtest.h"

/* Test null pointers */
TEST(CircleBuff, NullPtrWrite) {
  bfs::CircleBuf<uint8_t, 100> buff;
  std::size_t bytes_written = buff.Write(nullptr, 0);
  std::size_t bytes_read = buff.Read(nullptr, 0);
  EXPECT_EQ(0, bytes_written);
  EXPECT_EQ(0, bytes_read);
}
/* Test 0 length */
TEST(CircleBuff, ZeroLen) {
  bfs::CircleBuf<uint8_t, 100> buff;
  uint8_t data[0];
  std::size_t bytes_written = buff.Write(data, sizeof(data));
  std::size_t bytes_read = buff.Read(data, sizeof(data));
  EXPECT_EQ(0, bytes_written);
  EXPECT_EQ(0, bytes_read);
}
/* Test expected results */
TEST(CircleBuff, Expected) {
  bfs::CircleBuf<uint8_t, 10> buff;
  uint8_t data[5];
  EXPECT_EQ(10, buff.capacity());
  EXPECT_EQ(0, buff.size());
  std::size_t bytes_written = buff.Write(data, sizeof(data));
  EXPECT_EQ(5, bytes_written);
  EXPECT_EQ(10, buff.capacity());
  EXPECT_EQ(5, buff.size());
  uint8_t read_val[20];
  std::size_t bytes_read = buff.Read(read_val, 3);
  EXPECT_EQ(3, bytes_read);
  EXPECT_EQ(10, buff.capacity());
  EXPECT_EQ(2, buff.size());
  uint8_t data2[10];
  bytes_written = buff.Write(data2, sizeof(data2));
  EXPECT_EQ(8, bytes_written);
  EXPECT_EQ(10, buff.capacity());
  EXPECT_EQ(10, buff.size());
  bytes_read = buff.Read(read_val, 20);
  EXPECT_EQ(10, bytes_read);
  EXPECT_EQ(10, buff.capacity());
  EXPECT_EQ(0, buff.size());
  for (int i = 0; i < 10; i++) {
    buff.Write(data2[i]);
  }
  EXPECT_EQ(10, buff.capacity());
  EXPECT_EQ(10, buff.size());
  for (int i = 0; i < 5; i++) {
    uint8_t val;
    buff.Read(&val);
  }
  EXPECT_EQ(10, buff.capacity());
  EXPECT_EQ(5, buff.size());
  buff.Clear();
  EXPECT_EQ(10, buff.capacity());
  EXPECT_EQ(0, buff.size());
}


src/circle_buf.h
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2023 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#ifndef CIRCLE_BUF_SRC_CIRCLE_BUF_H_  // NOLINT
#define CIRCLE_BUF_SRC_CIRCLE_BUF_H_

#if defined(ARDUINO)
#include <Arduino.h>
#else
#include <cstring>
#include <cstddef>
#include <cstdint>
#include <algorithm>
#endif

namespace bfs {

template<typename T, size_t N>
class CircleBuf {
 public:
  bool Write(const T val) {
    if (capacity_ - size_) {
      buffer_[end_index_] = val;
      end_index_ = (end_index_ + 1) % capacity_;
      size_++;
      return true;
    }
    return false;
  }
  size_t Write(T const * const data, const size_t len) {
    if ((len == 0) || (!data)) return 0;
    #if defined(ARDUINO)
    vals_to_write_ = min(len, capacity_ - size_);
    #else
    vals_to_write_ = std::min(len, capacity_ - size_);
    #endif
    space_avail_ = capacity_ - end_index_;
    if (space_avail_ < vals_to_write_) {
      memcpy(buffer_ + end_index_, data, space_avail_ * sizeof(T));
      memcpy(buffer_, data + space_avail_,
            (vals_to_write_ - space_avail_) * sizeof(T));
      end_index_ = vals_to_write_ - space_avail_;
    } else {
      memcpy(buffer_ + end_index_, data, vals_to_write_ * sizeof(T));
      end_index_ += vals_to_write_;
    }
    size_ += vals_to_write_;
    return vals_to_write_;
  }
  bool Read(T * const data) {
    if (!data) {return false;}
    if (size_) {
      *data = buffer_[begin_index_];
      begin_index_ = (begin_index_ + 1) % capacity_;
      size_--;
    }
    return true;
  }
  size_t Read(T * const data, const size_t len) {
    if ((len == 0) || (!data)) return 0;
    #if defined(ARDUINO)
    vals_to_read_ = min(len, size_);
    #else
    vals_to_read_ = std::min(len, size_);
    #endif
    space_avail_ = capacity_ - begin_index_;
    if (space_avail_ < vals_to_read_) {
      memcpy(data, buffer_ + begin_index_, space_avail_ * sizeof(T));
      memcpy(data + space_avail_, buffer_,
            (vals_to_read_ - space_avail_) * sizeof(T));
      begin_index_ = vals_to_read_ - space_avail_;
    } else {
      memcpy(data, buffer_ + begin_index_, vals_to_read_ * sizeof(T));
      begin_index_ += vals_to_read_;
    }
    size_ -= vals_to_read_;
    return vals_to_read_;
  }
  inline size_t size() const {return size_;}
  inline size_t capacity() const {return capacity_;}
  void Clear() {
    begin_index_ = 0;
    end_index_ = 0;
    size_ = 0;
  }

 private:
  T buffer_[N];
  size_t begin_index_ = 0;
  size_t end_index_ = 0;
  size_t capacity_ = N;
  size_t size_ = 0;
  size_t vals_to_write_, vals_to_read_, space_avail_;
};

}  // namespace bfs

#endif  // CIRCLE_BUF_SRC_CIRCLE_BUF_H_ NOLINT


examples/circle_buf_ex.cc
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2020 Bolder Flight Systems
*/

#include "circle_buf/circle_buf.h"
#include <iostream>

int main() {
  CircularBuffer buff(10);
  uint8_t test[5] = {1, 2, 3, 4, 5};
  unsigned int ret = buff.Write(test, sizeof(test));
  std::cout << ret << std::endl;
  std::cout << std::endl;
  uint8_t read_val[20];
  ret = buff.Read(read_val, 3);
  std::cout << ret << std::endl;
  std::cout << std::endl;
  for (unsigned int i = 0; i < ret; i++) {
    std::cout << std::to_string(read_val[i]) << std::endl;
  }
  std::cout << std::endl;
  uint8_t test2[10] = {10, 20, 30, 40, 50, 60, 70, 80, 90, 100};
  ret = buff.Write(test2, sizeof(test2));
  std::cout << ret << std::endl;
  std::cout << std::endl;
  ret = buff.Read(read_val, 20);
  std::cout << ret << std::endl;
  for (unsigned int i = 0; i < ret; i++) {
    std::cout << std::to_string(read_val[i]) << std::endl;
  }
}


src/circle_buf/circle_buf.cc
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2020 Bolder Flight Systems
*/

#include "circle_buf/circle_buf.h"
#include <string.h>
#include <algorithm>

unsigned int CircularBuffer::Write(const std::vector<uint8_t> &data) {
  if (data.size() == 0) return 0;
  unsigned int bytes_to_write = std::min((unsigned int)data.size(), capacity_ - size_);
  unsigned int space_avail = capacity_ - end_index_;
  if (space_avail < bytes_to_write) {
    memcpy(buffer_.data() + end_index_, data.data(), space_avail);
    memcpy(buffer_.data(), data.data() + space_avail, bytes_to_write - space_avail);
    end_index_ = bytes_to_write - space_avail;
  } else {
    memcpy(buffer_.data() + end_index_, data.data(), bytes_to_write);
    end_index_ += bytes_to_write;
  }
  size_ += bytes_to_write;
  return bytes_to_write;
}

std::vector<uint8_t> CircularBuffer::Read(unsigned int bytes) {
  unsigned int bytes_to_read = std::min(bytes, size_);
  unsigned int space_avail = capacity_ - begin_index_;
  std::vector<uint8_t> ret(bytes_to_read);
  if (bytes == 0) return ret;
  if (space_avail < bytes_to_read) {
    memcpy(ret.data(), buffer_.data() + begin_index_, space_avail);
    memcpy(ret.data() + space_avail, buffer_.data(), bytes_to_read - space_avail);
    begin_index_ = bytes_to_read - space_avail;
  } else {
    memcpy(ret.data(), buffer_.data() + begin_index_, bytes_to_read);
    begin_index_ += bytes_to_read;
  }
  size_ -= bytes_to_read;
  return ret;
}

unsigned int CircularBuffer::Size() {
  return size_;
}

unsigned int CircularBuffer::Capacity() {
  return capacity_;
}


examples/cmake/circle_buf_example.cc
/*
* Brian R Taylor
* brian.taylor@bolderflight.com
* 
* Copyright (c) 2023 Bolder Flight Systems Inc
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the “Software”), to
* deal in the Software without restriction, including without limitation the
* rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
* sell copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
* IN THE SOFTWARE.
*/

#include "circle_buf.h"
#include <iostream>

int main() {
  /* Create a 10 byte circular buffer */
  bfs::CircleBuf<uint8_t, 10> buff;
  std::cout << "Starting Capacity: " << buff.capacity() << std::endl;
  std::cout << "Starting Size: " << buff.size() << std::endl << std::endl;
  /* Write 5 bytes to it */
  std::cout << "Writing 1 - 5 to buffer..." << std::endl;
  uint8_t test[] = {1, 2, 3, 4, 5};
  unsigned int ret = buff.Write(test, sizeof(test));
  std::cout << "Bytes written: " << ret << std::endl;
  std::cout << "New Capacity: " << buff.capacity() << std::endl;
  std::cout << "New Size: " << buff.size() << std::endl;
  std::cout << std::endl;
  /* Read 3 bytes from buffer */
  std::cout << "Reading 3 bytes from buffer..." << std::endl;
  uint8_t read_val[20];
  ret = buff.Read(read_val, 3);
  std::cout << "Bytes read " << ret << std::endl;
  std::cout << "Bytes: " << std::endl;
  for (unsigned int i = 0; i < ret; i++) {
    std::cout << std::to_string(read_val[i]) << std::endl;
  }
  std::cout << std::endl;
  /* Try to write 10 bytes to buffer, only space for 8 */
  std::cout << "Writing 10 - 100 to buffer..." << std::endl;
  uint8_t test2[] = {10, 20, 30, 40, 50, 60, 70, 80, 90, 100};
  ret = buff.Write(test2, sizeof(test2));
  std::cout << "Bytes written: " << ret << std::endl;
  std::cout << "New Capacity: " << buff.capacity() << std::endl;
  std::cout << "New Size: " << buff.size() << std::endl;
  std::cout << std::endl;
  /* Try to read 20 bytes from buffer, only 10 stored */
  std::cout << "Reading 20 bytes from buffer..." << std::endl;
  ret = buff.Read(read_val, sizeof(read_val));
  std::cout << "Bytes read " << ret << std::endl;
  std::cout << "Bytes: " << std::endl;
  for (unsigned int i = 0; i < ret; i++) {
    std::cout << std::to_string(read_val[i]) << std::endl;
  }
  std::cout << std::endl;
  /* Write 10 bytes to the buffer */
  std::cout << "Writing 10 - 100 to buffer, one at a time..." << std::endl;
  for (unsigned int i = 0; i < sizeof(test2); i++) {
    ret = buff.Write(test2[i]);
  }
  std::cout << "New Capacity: " << buff.capacity() << std::endl;
  std::cout << "New Size: " << buff.size() << std::endl;
  std::cout << std::endl;
  /* Read 5 bytes from the buffer */
  std::cout << "Reading 5 bytes from the buffer, one at a time..." << std::endl;
  std::cout << "Bytes: " << std::endl;
  for (unsigned int i = 0; i < 5; i++) {
    uint8_t val;
    bool ret = buff.Read(&val);
    if (ret) {
      std::cout << std::to_string(val) << std::endl;
    }
  }
  std::cout << "New Capacity: " << buff.capacity() << std::endl;
  std::cout << "New Size: " << buff.size() << std::endl;
  std::cout << std::endl;
  /* Clearing the buffer */
  std::cout << "Clearing the buffer..." << std::endl;
  buff.Clear();
  std::cout << "New Capacity: " << buff.capacity() << std::endl;
  std::cout << "New Size: " << buff.size() << std::endl; 
  std::cout << std::endl; 
}
